### ISSUE TITLE
MI 4.2.0 - SAML integration support

### ISSUE THREAD

**Issue by sarindufit** [wso2/api-manager#1325]
### Problem

This git issue is related to https://github.com/wso2-enterprise/wso2-apim-internal/issues/1678.

MI 4.2.0 has not been properly tested for SAML related use cases. As a result, multiple issues are observed when implemented SAML 2.0 integrations. The said issues are documented in the above ticket. 

### Solution

Improve MI 4.2.0 to support SAML related use cases. 

### Affected Component

MI

### Version

4.2.0-Beta

**Comment by rosensilva** [comment]
Hi @sarindufit
I have debugged the SAML flow with MI 4.2.0 pre-beta pack and the following points were observed
* According to the stack trace ``` [2023-01-25 14:25:45,020] ERROR {NativeWorkerPool} - Uncaught exception java.lang.NoClassDefFoundError: org/apache/xml/security/exceptions/XMLSecurityException ```  org/apache/xml/security/exceptions package is not visible to the custom mediator with saml implementation
* We have packed this class as a bundle as  ```<bundleDef>org.wso2.santuario.orbit:xmlsec</bundleDef>``` in org.wso2.micro.integrator.identity.xacml.mediator.feature
* When debugging our orbit bundle for opensaml/3.3.1.wso2v7 we can see that xmlsec dependency is used but not packed.
* After adding it as an embedded dependency the SAML scenario worked as expected
* Please test by adding the jar inside /wso2mi-4.2.0-alpha/wso2/components/plugins/opensaml_3.3.1.wso2v7.jar



[patch.zip](https://github.com/wso2/api-manager/files/10553147/patch.zip)

**Comment by sarindufit** [comment]
Hi Rosen, 

Please note that I can confirm that the below issue is now resolved.
```
[2023-01-25 14:25:45,020] ERROR {NativeWorkerPool} - Uncaught exception java.lang.NoClassDefFoundError: org/apache/xml/security/exceptions/XMLSecurityException
```

However, the execution flow now results in the below error. Note that the ws-security header is not sent out to the backend. I have attached the wire log for your reference. 

```
[2023-02-01 12:48:40,186] ERROR {org.apache.synapse.core.axis2.Axis2Sender} - {proxy:ch_suva_wsx_monitoring_platform_sap_crm_saml_v1} Unexpected error during sending message out. Error in signature with X509Token org.apache.axis2.AxisFault: Error in signature with X509Token
	at org.apache.rampart.handler.RampartSender.invoke(RampartSender.java:76)
	at org.apache.axis2.engine.Phase.invokeHandler(Phase.java:340)
	at org.apache.axis2.engine.Phase.invoke(Phase.java:313)
	at org.apache.axis2.engine.AxisEngine.invoke(AxisEngine.java:261)
	at org.apache.axis2.engine.AxisEngine.send(AxisEngine.java:426)
	at org.apache.synapse.core.axis2.DynamicAxisOperation$DynamicOperationClient.send(DynamicAxisOperation.java:206)
	at org.apache.synapse.core.axis2.DynamicAxisOperation$DynamicOperationClient.executeImpl(DynamicAxisOperation.java:188)
	at org.apache.axis2.client.OperationClient.execute(OperationClient.java:149)
	at org.apache.synapse.core.axis2.Axis2FlexibleMEPClient.send(Axis2FlexibleMEPClient.java:675)
	at org.apache.synapse.core.axis2.Axis2Sender.sendOn(Axis2Sender.java:86)
	at org.apache.synapse.core.axis2.Axis2SynapseEnvironment.send(Axis2SynapseEnvironment.java:591)
	at org.apache.synapse.endpoints.AbstractEndpoint.send(AbstractEndpoint.java:423)
	at org.apache.synapse.endpoints.AddressEndpoint.send(AddressEndpoint.java:79)
	at org.apache.synapse.core.axis2.ProxyServiceMessageReceiver.receive(ProxyServiceMessageReceiver.java:261)
	at org.apache.axis2.engine.AxisEngine.receive(AxisEngine.java:180)
	at org.apache.synapse.transport.passthru.ServerWorker.processEntityEnclosingRequest(ServerWorker.java:472)
	at org.apache.synapse.transport.passthru.ServerWorker.run(ServerWorker.java:188)
	at org.apache.axis2.transport.base.threads.NativeWorkerPool$1.run(NativeWorkerPool.java:172)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.rampart.RampartException: Error in signature with X509Token
	at org.apache.rampart.builder.AsymmetricBindingBuilder.doSignature(AsymmetricBindingBuilder.java:789)
	at org.apache.rampart.builder.AsymmetricBindingBuilder.doSignBeforeEncrypt(AsymmetricBindingBuilder.java:457)
	at org.apache.rampart.builder.AsymmetricBindingBuilder.build(AsymmetricBindingBuilder.java:97)
	at org.apache.rampart.MessageBuilder.build(MessageBuilder.java:147)
	at org.apache.rampart.handler.RampartSender.invoke(RampartSender.java:65)
	... 20 more
Caused by: org.apache.ws.security.WSSecurityException: Signature creation failed (Cannot setup signature data structure); nested exception is: 
	org.apache.xml.security.transforms.TransformationException: Unknown transformation. No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#STR-Transform
Original Exception was org.apache.xml.security.transforms.InvalidTransformException: Unknown transformation. No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#STR-Transform
	at org.apache.ws.security.message.WSSecSignature.addReferencesToSign(WSSecSignature.java:640)
	at org.apache.rampart.builder.AsymmetricBindingBuilder.doSignature(AsymmetricBindingBuilder.java:779)
	... 24 more
Caused by: org.apache.xml.security.transforms.TransformationException: Unknown transformation. No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#STR-Transform
Original Exception was org.apache.xml.security.transforms.InvalidTransformException: Unknown transformation. No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#STR-Transform
	at org.apache.xml.security.transforms.Transforms.addTransform(Transforms.java:197)
	at org.apache.ws.security.message.WSSecSignature.addReferencesToSign(WSSecSignature.java:561)
	... 25 more
Caused by: org.apache.xml.security.transforms.InvalidTransformException: Unknown transformation. No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#STR-Transform
	at org.apache.xml.security.transforms.Transform.initializeTransform(Transform.java:336)
	at org.apache.xml.security.transforms.Transform.<init>(Transform.java:117)
	at org.apache.xml.security.transforms.Transforms.addTransform(Transforms.java:193)
	... 26 more
```

**When I tested the same implementation with wso2mi-4.1.0, this issue is not apparent. Therefore, I believe mi 4.2.0 still needs further changes to the SAML integration flow** 

Thanks,
Sarindu.

**Comment by sarindufit** [comment]
[wso2carbon.log](https://github.com/wso2/api-manager/files/10554404/wso2carbon.log)

**Comment by sarindufit** [comment]
Hi Rosen, 

As per your request, I'm attaching the artifacts which you can use to test this issue. Please copy the jar file to the lib directory and the car file to the carbonapps directory. 

Upon deployment, you can use the below WSDL to create a SOAP request in SOAP UI. 

http://localhost:8290/services/ch_suva_wsx_monitoring_platform_sap_crm_saml_v1?wsdl

[artifacts.zip](https://github.com/wso2/api-manager/files/10673677/artifacts.zip)

Thanks,
Sarindu.

**Comment by rosensilva** [comment]
Hi All,

I have debugged the use case with SAML headers and SOAP requests. The following points were discovered during the debugging.

* The said security fix for the opensaml orbit bundle had issues. We have migrated opensaml_3.3.1.wso2v4 old mi 4.1.0 to 
opensaml_3.3.1.wso2v7 in new mi 4.2.0. inside the orbit bundle, there were several dependency updates due to security vulnerabilities. 
* By further analyzing the opensaml orbit bundles I noticed that the base opensaml update was not causing the issue, but the dependency version limitation of xml.security package causes the class not to find the issue.
* That fix was also done for a security fix from the IAM team. Along the way, we folked the xmlsec from apache and ported the security fix, hence we were using the forked version of xmlsec.
* When foking the xmlsec from apache we used the xmlsec version 1.5.6 instead of the 1.5.4 we had originally with the mi 4.1.0
* This minor version update causes the 
```
 Caused by: org.apache.xml.security.transforms.InvalidTransformException: Unknown transformation.
 No handler installed for URI http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security
-1.0#STR-Transform
```
error and was able to get it resolved once I downgrade the xmlsec version to 1.5.4.
* Hence we can fix this problem by downgrading the xmlsec base version for the forked wso2-xmlsec repo.
* We also need to release a new opensaml orbit bundle by increasing the lower version range for the xmlsec library.

We will update the progress of this ticket.
Adding @GDLMadushanka as he has done the initial folk for the xml-sec and currently doing the base version change.
